### DISCUSSION TITLE
Consulta atributo JSON

### DISCUSSION THREAD

**Discussion by nubevi** [dds-jv/foro#253] [TPA - 2021] [answered]
Buenas tardes,

Estaba modelando los datos del siguiente JSON al ejecutar GET y veo que tiene como nombre "long". El problema es que el IDE  me lo reconoce como palabra reservada. Hay alguna forma de poder recibir ese parámetro cambiando el nombre en mi codigo?

![image](https://user-images.githubusercontent.com/26943847/120082187-ee90e500-c097-11eb-95f5-0fcd4f1699c5.png)

**Comment by nanderson87** [comment, marked as answer]
Hola @nubevi 

La respuesta corta es que si. Se puede hacer de muchas maneras. A continuación te dejo unos ejemplos.

Ejemplo 1: `jersey`

Response body de ejemplo
```
{
    "long": -34,
    "lat": 0,
    "nombre": "Casa A"
}
```

Clase que modela la respuesta de la API
```
public class ResponseBody {
   private Integer longitud;
   private Integer latitud;
   private String nombreHogar;
   
   //Constructor(es) - getter(s) - etc
}
```

Creación del cliente jersey
```
Client client = Client.create();
```

Uso del cliente jersey
```
//Armado de la request
ClientResponse response = builder.get(ClientResponse.class);
.. 
HasMap<String, Object> responseBodyMap = response.getEntity(Map.class);
//Una vez tengamos el mapa se puede construir nuestro objeto `ResponseBody` a partir de este extrayendo cada key.

```
Esto sería una solución más bien manual. Hay muchos frameworks que nos ayudan a trabajar con los JSON como objetos o "mappear"  un JSON a un objeto y viceversa. Alguno de estos frameworks para java son  [jackson](https://github.com/FasterXML/jackson ) o [gson](https://github.com/google/gson).

Muchos de los cliente HTTP tienen su propio mappear o nos da la facilidad de trabajar con alguno de estos frameworks ya que es algo bastante común (y jersey no es la excepción) 

Ejemplo 2: `jersey-jackson`


Mismo cuerpo de respuesta
```
{
    "long": -34,
    "lat": 0,
    "nombre": "Casa A"
}
```
pom.xml
```
  <dependency>
      <groupId>com.sun.jersey</groupId>
      <artifactId>jersey-client</artifactId>
      <version>1.19</version>
  </dependency>
  <dependency>
      <groupId>com.sun.jersey</groupId>
      <artifactId>jersey-json</artifactId>
      <version>1.19</version>
 </dependency>
```


```
public class ResponseBody {

   //Esta anotación le va a servir a jackson para saber como mapear el JSON a nuestro objeto.
   //En este caso le estamos diciendo que el field con nombre long del JSON tiene que terminar siendo
   //nuestro atributo longitud en la clase ResponseBody
   @JsonProperty("long")  
   private Integer longitud;
   
   @JsonProperty("lat")
   private Integer latitud;
   
   @JsonProperty("nombre")
   private String nombreHogar;
   
   public ResponseBody() { 
   }
   
   //Constructor(es) - getter(s) - etc
}
```

Creación del cliente jersey
```
ClientConfig clientConfig = new DefaultClientConfig();
clientConfig.getFeatures().put(JSONConfiguration.FEATURE_POJO_MAPPING, Boolean.TRUE);
Client client = Client.create(clientConfig);
```

Uso del cliente jersey
```
ClientResponse response = builder.get(ClientResponse.class);
ResponseBody responseBody = response.getEntity(ResponseBody.class);
```
En este caso directamente podemos convertir la respuesta representada en un JSON al objeto que modelamos.


Dependiendo  del cliente HTTP, la versión del mismo, si usas un biblioteca/framework de conversión de JSON a Objecto - Objeto a JSON o no van a variar algunas cosas pero la idea es la misma.

**Reply by nubevi** [reply]
Gracias Nico!! :)
